torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -53,3 +70,96 @@ def _auto_wrap(
         )
         auto_wrap_kwargs["auto_wrap_policy"] = auto_wrap_policy
     _recursive_wrap(**auto_wrap_kwargs, **fsdp_kwargs)
+
+
+def _get_submodule_to_states(
+    root_module: nn.Module,
+    auto_wrap_policy: Callable,
+    ignored_modules: Set[nn.Module],
+    ignored_params: Set[nn.Parameter],
+) -> Dict[nn.Module, SubmoduleState]:
+    """
+    Returns a mapping from submodule to its parameters, buffers, parameter
+    names, and buffer names, where each entry logically represents a grouping
+    according to the given auto wrap policy and ignored modules/parameters.
+    However, this method does not actually perform any module wrapping.
+
+    The mapped-to values are the states from the subtree rooted at the
+    corresponding submodule key, excluding child submodules in the mapping and
+    ignored state. Sibling submodules cannot be grouped together. The parameter
+    and buffer names are prefixed starting from the submodule.
+
+    Each non-ignored parameter and buffer appears exactly once in the returned
+    ``dict``, and the ``dict`` is ordered by increasing tree depth. A mapped-to
+    parameter list may be empty if the submodule has no parameters or if its
+    parameters were assigned to a parent submodule instead.
+    """
+    # Record the modules to wrap without actually wrapping
+    wrapped_modules: List[nn.Module] = []  # these are only logically wrapped
```

Comment:
no need to change anything. JFYI, existing tracing techniques usually have issues on tracing inplace tensor functions and they have to do special-handling to workaround. There is a discussion whether we should provide full functional ops and can do another optimization pass to re-inplace ops. It might worth considering whether we would like to have similar functional design in our APIs as well.